state.go

```diff
@@ -557,6 +560,98 @@ func (s *State) Channel(channelID string) (*Channel, error) {
 	return nil, ErrStateNotFound
 }
 
+// ThreadMemberUpdate updates the ThreadMember object for the user,
+// in a specific thread.
+func (s *State) ThreadMemberUpdate(t *ThreadMember) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	channel, ok := s.channelMap[t.ID]
+	if ok {
+		s.Lock()
+		defer s.Unlock()
+
+		channel.Member = t
+		return nil
+	}
+
+	return ErrStateNotFound
+}
+
+// ThreadMembersUpdate updates the member object of the thread.
+// If the user was removed,  the ThreadMember object in the state
+// will be set to nil.
+func (s *State) ThreadMembersUpdate(t *ThreadMembersUpdate) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	thread, ok := s.channelMap[t.ID]
+	if ok {
+		for _, memberID := range t.RemovedMembersIDs {
+			if memberID == s.User.ID {
+				thread.Member = nil
+				break
+			}
+		}
+	}
+
+	return ErrStateNotFound
+}
+
+// ThreadListSync syncs new threads and threads member that
+// the bot gained access to.
+// https://discord.com/developers/docs/topics/gateway#thread-list-sync
+func (s *State) ThreadListSync(t *ThreadListSync) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	g, err := s.Guild(t.GuildID)
+	if err != nil {
+		return err
+	}
+
+	if len(t.ChannelIDs) > 0 {
+		s.Lock()
+		defer s.Unlock()
+
+		for _, parentChannelID := range t.ChannelIDs {
+			for _, stateChannel := range s.channelMap {
+				if stateChannel.IsThread() && stateChannel.ParentID == parentChannelID && !stateChannel.ThreadMetadata.Archived {
+					for i, c := range g.Channels {
+						if c.ID == stateChannel.ID {
+							g.Channels = append(g.Channels[:i], g.Channels[i+1:]...)
+							break
+						}
+					}
+					delete(s.channelMap, stateChannel.ID)
+				}
+			}
+		}
+
+		for _, channel := range t.Threads {
+			s.channelMap[channel.ID] = channel
+			g.Channels = append(g.Channels, channel)
+		}
+
+		for _, member := range t.Members {
+			channel, ok := s.channelMap[member.ID]
+			if ok {
+				channel.Member = member
+			}
+		}
+
+		return nil
+	}
+
+	return ErrStateNotFound
+}
```

Comment:
This appears to be copied from https://github.com/bwmarrin/discordgo/pull/982/files/ecd5d0d76e298de23c533d7e4f1f8c026930fb07#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aR605-R659, and retains the same very unusual structure of the nested loops (it's so non-straightforward that I cannot imagine two people coming to this independently), the same misunderstanding of what the case of `len(t.ChannelIDs) == 0` means, and the same return of `ErrStateNotFound` when that error _really_ doesn't make a ton of sense in this context.